model/atmosphere/dycore/src/icon4py/model/atmosphere/dycore/stencils/compute_edge_diagnostics_for_velocity_advection.py

```diff
@@ -181,46 +74,35 @@ def _compute_derived_horizontal_winds_and_ke_and_horizontal_advection_of_w_and_c
     fa.EdgeKField[ta.vpfloat],
     fa.EdgeKField[ta.vpfloat],
 ]:
-    (
-        tangential_wind,
-        vn_on_half_levels,
-        tangential_wind_on_half_levels,
-        horizontal_kinetic_energy_at_edges_on_model_levels,
-        contravariant_correction_at_edges_on_model_levels,
-    ) = _compute_derived_horizontal_winds_and_kinetic_energy_and_contravariant_correction(
-        vn,
-        rbf_vec_coeff_e,
-        wgtfac_e,
-        ddxn_z_full,
-        ddxt_z_full,
-        contravariant_correction_at_edges_on_model_levels,
-        nflatlev,
-        tangential_wind_on_half_levels,
-        tangential_wind,
-        vn_on_half_levels,
-        horizontal_kinetic_energy_at_edges_on_model_levels,
-        k,
-        nlev,
-        skip_compute_predictor_vertical_advection,
+    tangential_wind = _compute_tangential_wind(vn, rbf_vec_coeff_e)
+    horizontal_kinetic_energy_at_edges_on_model_levels = _compute_horizontal_kinetic_energy(
+        vn, tangential_wind
     )
+    vn_on_half_levels = _interpolate_to_half_levels(k, wgtfac_e, vn)
 
-    k = broadcast(k, (dims.EdgeDim, dims.KDim))
+    contravariant_correction_at_edges_on_model_levels = where(
+        nflatlev <= k,
+        _compute_contravariant_correction(vn, ddxn_z_full, ddxt_z_full, tangential_wind),
+        contravariant_correction_at_edges_on_model_levels,
+    )
 
+    # TODO(havogt): for now and performance, we should split this into two programs for
```

Comment:
This might not be necessary?
This conditional is known at JIT time, which means this will be optimized away automatically by a dead code elimination pass.
Furthermore, the workflow of the DaCe team for optimizing the stencils is to look at all relevant combinations of JIT time known booleans and then to optimize all of those stencils separately.